### DISCUSSION TITLE
Accessing the repeat variable to make multiple files/dirs etc.

### DISCUSSION THREAD

**Discussion by Joyje** [gokcehan/lf#1180] [Q&A] [answered]
If I press "j" I will move my selection down one step. If I instead press "12j" I will move my selection down 12 steps.

I want to utilize this when making files and directories. I have the following in my lfrc:
##################################################################

`cmd touch %{{`
`    printf "Name New File: "`
`    read latestfile`
`    touch ${latestfile}`
`}}`

...

`map i touch`

#################################################################

Now pressing "i" will make a new file, however pressing "3i" won't make 3 files. Only one file is touched.
How can I access the "3" part of the command so I can change the last line to:

    touch $(seq -f "${latestfile}%g" <insert repeat variable here>)

UPDATE 1

Working flawlessly so far using the solution provided by @joelim-work! Absolutely love it!

I changed my bash function to:

`cmd newfile %{{`
`    printf "Name new file(s): "`
`    read newfile`
`    if [ $lf_count -gt 1 ]; then`
`        for file in ${newfile}; do`
`            touch $(seq -f "${file}%g" ${lf_count})`
`        done`
`    else`
`        touch ${newfile}`
`    fi`
`}}`

Changed the name from "touch" to "newfile" as to not mix up my own command with the shell command.

One of my favorites so far is this one however:

`cmd append %{{`
`    printf "Append to selected file(s): "`
`    read appendage`
`    for file in $fx; do`
`        for count in 'seq $lf_count'; do` *
`            echo ${appendage} >> ${file}`
`        done`
`    done`
`}}`
*(replace the ' signs with " `` " signs)

Couldn't be happier with the results!
Now the only question is why isn't this already built in to `lf`? It improves everything!

UPDATE 2

The code has been pushed to the repo! Huge thanks to @joelim-work & @gokcehan!  Expect this to work out of the box in v29 and onward.

**Comment by joelim-work** [comment]
When `lf` runs a shell command, it exports certain variables such as the selected files, terminal width/height and user settings:

https://github.com/gokcehan/lf/blob/6ffc4a4c8aef3cf84b04b44515240612865a4e4e/app.go#L490-L492

Currently, the command count is not exported, so your command won't work.

---

An alternative approach is to define the `touch` command to accept arguments (filename and count) and forward them to the shell invocation:

```
cmd touch $touch $(seq -f "$1%g" "$2")

map i push :touch<space>
```

You can type `:touch foo 10` to create files from `foo1` to `foo10`, and press `i` to fill in the `:touch ` part for you.

**Reply by Joyje** [reply]
There must be some way to register keystrokes in lf to export them as a variable. Maybe it will be exported as a string with both count and command, but that won't be an issue since we could simply cut of anything beyond the count, save it as a variable and use that variable in our own commands. But do you seriously mean that I need to install a key logger and run it in the background every time lf is running in order to do this? Aren't there some sort of input log where lf stores the latest inputs it received?

I mean, what is the point of having a command count if we can't use it? Seems quite redundant to make every single shell function demand a manual input for command count when lf already gives us the option.

**Reply by joelim-work** [reply]
The reason I suggested the above approach was that it involved practically the same amount of effort (press `i` then `foo 10`, compared to `10i` then `foo`) but without having to make any code changes to `lf`.

If you absolutely insist that `lf` should export the command count as a variable, then you would have to modify the code to do so:

```diff
--- a/eval.go
+++ b/eval.go
@@ -1254,20 +1254,21 @@ func insert(app *app, arg string) {
                }
                fallthrough
        default:
                app.ui.menuBuf = nil
                app.menuCompActive = false
                app.ui.cmdAccLeft = append(app.ui.cmdAccLeft, []rune(arg)...)
        }
 }

 func (e *callExpr) eval(app *app, args []string) {
+       os.Setenv("lf_count", strconv.Itoa(e.count))
        switch e.name {
        case "up":
                if !app.nav.init {
                        return
                }
                if app.nav.up(e.count) {
                        app.ui.loadFile(app, true)
                        app.ui.loadFileInfo(app.nav)
                }
        case "half-up":
```

With this change you should be able to use `lf_count` as a variable:

```
cmd touch %{{
    printf "Name New File: "
    read latestfile
    touch $(seq -f "${latestfile}%g" "$lf_count")
}}

map i touch
```

**Reply by Joyje** [reply]
Whoa! I didn't expect such a great answer! Huge thanks for that! I will try it out with my setup and let you know how it went.

The first solution you suggested would cause issues if I wanted to create multiple files with separate names. I would prefer if pressing `i` and typing `foo 10` makes 2 files. One named "foo" and one named "10". This seems like a much better practice, so it makes more sense to use the command count for these situations.

**Reply by Joyje** [reply]
Working flawlessly so far! Absolutely love it!

I changed my bash function to:

`cmd newfile %{{`
`    printf "Name new file(s): "`
`    read newfile`
`    if [ $lf_count -gt 1 ]; then`
`        for file in ${newfile}; do`
`            touch $(seq -f "${file}%g" ${lf_count})`
`        done`
`    else`
`        touch ${newfile}`
`    fi`
`}}`

Changed the name from "touch" to "newfile" as to not mix up my own command with the shell command.

One of my favorites so far is this one however:

`cmd append %{{`
`    printf "Append to selected file(s): "`
`    read appendage`
`    for file in $fx; do`
`        for count in 'seq $lf_count'; do`
`            echo ${appendage} >> ${file}`
`        done`
`    done`
`}}`
(replace the ' signs with " `` " signs)
Couldn't be happier with the results!

Now the only question is why isn't this already built in to lf? You think you could push your code to the repo?
It only gives a slight error when starting lf that the variable is empty, but it doesn't break anything as far as I can tell. Instead it improves everything!

**Reply by joelim-work** [reply]
No worries, I didn't realise you wanted to conceptually separate the count (sort of like a modifier) from the regular arguments.

The reason why the `lf` doesn't export the count is simply because up until now, nobody has ever mentioned such a use case (this is a typical scenario). This does fall under the category of a 'new feature' though - while I cannot 100% guarantee that it will be accepted, I can see the potential uses for it.

Incidentally Vimscript provides a `v:count` variable, which is quite similar:

```vim
function! EchoCount()
    echo v:count
endfunction

nnoremap <silent> e :<c-u>call EchoCount()<cr>

" Typing '123e' will result in printing '123'
```

Anyway please test the patch, if it works for you then I wouldn't mind submitting a pull request on your behalf.

**Reply by Joyje** [reply]
I realized you should probably copy your comment and paste it as an answer so I can mark this as answered. My latest comment should probably be an update also rather than a comment. I will paste it as an update after I can mark yours as an answer. Thanks a lot for your solution! I really appreciate it!

**Reply by joelim-work** [reply]
Oh I didn't see your comment just now, about the error where the variable is empty when starting `lf`, how are you managing to trigger it? The count should default to 1, so you must be somehow invoking your script without hitting the line of code that exports `lf_count`.

**Reply by Joyje** [reply]
Right at the startup I have an error message at the bottom of the screen saying `number: value should be empty, 'true', or 'false'`

**Reply by joelim-work** [reply]
Could you please provide your `lfrc` file? I suspect it might be a separate issue and you probably have something invalid like `set number off`.

**Reply by Joyje** [reply]
Oh! I have `set number on`. Thats probably it. I used v27 before I used your code. Since I had to build from source I decided to update to v28 while I was at it. Didn't notice that the numbered lines went away with the update until you mentioned it now as a separate issue, so I assumed it was due to your code. `on` worked in v27, but v28 expects `'true'` instead apparently.

This is even better since that means there are no errors so far due to your line of code. Great!

**Reply by joelim-work** [reply]
The syntax for setting boolean options was modified very recently in 85512625f87621ca484859f88d72ba0df0739d63, you can read more about it in #758

**Reply by gokcehan** [reply]
@joelim-work @Joyje I think this was requested before in #62 . If the patch is somehow useful and working as expected, feel free to send it as a PR.

**Comment by joelim-work** [comment, marked as answer]
This behaviour is currently not supported in `lf`, because up until now nobody has ever mentioned such a use case.

In order to achieve this, the code would have to be modified to export the command count as a variable:

```diff
--- a/eval.go
+++ b/eval.go
@@ -1254,20 +1254,21 @@ func insert(app *app, arg string) {
                }
                fallthrough
        default:
                app.ui.menuBuf = nil
                app.menuCompActive = false
                app.ui.cmdAccLeft = append(app.ui.cmdAccLeft, []rune(arg)...)
        }
 }

 func (e *callExpr) eval(app *app, args []string) {
+       os.Setenv("lf_count", strconv.Itoa(e.count))
        switch e.name {
        case "up":
                if !app.nav.init {
                        return
                }
                if app.nav.up(e.count) {
                        app.ui.loadFile(app, true)
                        app.ui.loadFileInfo(app.nav)
                }
        case "half-up":
```

With this change you should be able to use `lf_count` as a variable:

```
cmd touch %{{
    printf "Name New File: "
    read latestfile
    touch $(seq -f "${latestfile}%g" "$lf_count")
}}

map i touch
```
